src/engine.ml

```diff
@@ -1690,167 +1697,74 @@ let incoming ?(is_not_taken = fun _ip -> false) state buf =
-                  let decrypted =
-                    Aes_ctr.decrypt ~key:tls_crypt.their_key ~ctr encrypted
-                  in
-                  let* ack =
-                    Packet.Tls_crypt.decode_decrypted_ack cleartext decrypted
-                  in
-                  let p = `Ack ack in
-                  let to_be_signed = Packet.Tls_crypt.to_be_signed key p in
-                  let computed_hmac =
-                    Mirage_crypto.Hash.SHA256.hmac ~key:tls_crypt.their_hmac
-                      to_be_signed
-                  in
-                  let* () =
-                    (* XXX maybe just ignore? *)
-                    if Eqaf_cstruct.equal computed_hmac cleartext.hmac then
-                      Ok ()
-                    else Error (`Bad_mac (state, computed_hmac, (key, p)))
-                  in
-                  match expected_packet state.session ch.transport p with
-                  | Error e ->
-                      (* XXX: only in udp mode? *)
-                      Log.warn (fun m -> m "ignoring bad packet %a" pp_error e);
-                      Ok (state, out, None)
-                  | Ok (session, transport) ->
-                      let state = { state with session }
-                      and ch = { ch with transport } in
-                      Ok (set_ch state ch, out, act))
-              | Packet.Hard_reset_client_v3, Client_tls_crypt _ ->
-                  Error (`No_transition (ch, op, payload))
-              | _control_op, Client_tls_crypt { tls_crypt = tls_crypt, wkc; _ }
-                -> (
-                  let* cleartext, off =
+                              (state, out, est @ acts))))
+              | _, Client_tls_crypt { tls_crypt = tls_crypt, wkc_opt; _ } -> (
+                  let* cleartext, encrypted =
                     Packet.Tls_crypt.decode_cleartext_header payload
                   in
                   let module Aes_ctr = Mirage_crypto.Cipher_block.AES.CTR in
-                  let encrypted = Cstruct.shift payload off in
                   let iv = Cstruct.sub cleartext.hmac 0 16 in
                   let ctr = Aes_ctr.ctr_of_cstruct iv in
                   let decrypted =
                     Aes_ctr.decrypt ~key:tls_crypt.their_key ~ctr encrypted
                   in
-                  let* ((_, _, data) as control) =
-                    Packet.Tls_crypt.decode_decrypted_control cleartext
-                      decrypted
+                  let* p =
+                    Packet.Tls_crypt.decode_decrypted_ack_or_control cleartext
+                      op decrypted
                   in
                   let* needs_wkc =
-                    match op with
-                    | Hard_reset_server_v2 ->
+                    match p with
+                    | `Control (Packet.Hard_reset_server_v2, (_, _, data)) ->
                         Packet.decode_early_negotiation_tlvs data
                     | _ -> Ok false
                   in
```

Comment:
A downside is we try to decode the early negotiation TLVs even for tls-crypt. We could match on `wkc_op, op` and only do the decoding if the wkc is present (tls-crypt-v2).